### PR TITLE
Code improvements

### LINKED ISSUE #57: Add try except blocks to python code
I have mostly relied on egpg to provide correct data to python bindings.
Like the file passed always exists and key passed is correct etc.

At a later point we need to check this in python scripts also. Like consider a case when file is deleted between checking by egpg and then its use by python script.

Also expectations from gpgme bindings are not well documented and I don't know which may pop up. This is important aspect that will be revealed while exhaustive testing. For now I am just focussing on getting pgpg working inside the container and just checking for a few basic tests.

### LINKED ISSUE #65: Simplify code for verify
as discussed in #61 use the following for simplifying verify code
```
.
.
.
               '''.format(user=c.get_key(signature.fpr).uids[0].uid,
                             fingerprint=signature.fpr,
                             time=time.ctime(signature.timestamp))
```

### LINKED ISSUE #79: Make code consistant
- [x] no brackets around if
- [x] pep8

### LINKED ISSUE #83: Reorganize error messages
I think that the error messages like this:
https://github.com/EasyGnuPG/pgpg/blob/gnupg-2.2/src/cmd/open.sh#L29-L30
should be printed from the python code.

### PR DESCRIPTION
**Pull request by diveshuttam**
This closes #83, closes #65, closes #79 and closes #57 .

### CODE REVIEW

**Comment by diveshuttam** [comment]
This **needs** to be based on #81. I am closing this for now, will open again after #81 is completed.

File: src/gpg/contact/delete.py [hunk 1]
```diff
@@ -21,10 +21,11 @@ def delete(contacts, force):
                 if ans.lower() == 'y' or force:
                     c.op_delete(key, False)
 
-    except BaseException:
+    except gpg.errors.GpgError as e:
         if os.environ['DEBUG'] == 'yes':
             raise
-        exit(2)
+        print(e, file=sys.stderr, flush=True)
+        exit(1)
```
Comment (dashohoxha) [inline src/gpg/contact/delete.py]: These 4 lines seem to be the same everywhere. Why don't you create a utility (auxiliary) function to encapsulate them, for example: `handle_exception(e)`

File: src/gpg/contact/export.py [hunk 2]
```diff
@@ -6,11 +6,16 @@
...
     else:
-        export_file = open(export_path, "w")
-
+        try:
+            export_file = open(export_path, "w")
+        except PermissionError:
+            print("Not enough permissions to write to {path}".format(
+                path=export_path
+            ), file=sys.stderr, flush=True)
+            exit(1)
```
Comment (dashohoxha) [inline src/gpg/contact/export.py]: These lines seem to be the same everywhere. Why don't you simplify them with a utility function like this: 
```
print_error("Not enough permissions to write to {path}".format(path=export_path))
```

File: src/gpg/contact/delete.py (hunk 1 above)
Comment (sourabhtk37) [reply src/gpg/contact/delete.py]: Also shouldn't the print statement be in the else statement?

Comment (diveshuttam) [reply src/gpg/contact/delete.py]: Umm I guess raise causes an exit as well.

File: src/gpg/fn/interact.py [hunk 3]
```diff
@@ -14,18 +14,18 @@ def edit_fnc(self, status, args, out=None):
...
+        if status == "PINENTRY_LAUNCHED":
+            cmd = None
+        else:
             cmd = self.cmds[self.step]
             self.step += 1
-            if self.step == len(self.cmds):
-                self.done = True
-        else:
-            cmd = None
+            self.done = len(self.cmds) == self.step
```
Comment (sourabhtk37) [inline src/gpg/fn/interact.py]: This doesn't seem that readable. Maybe try ternary operation for one-liner:
```python 
True if <expr> else False

File: src/gpg/fn/print_key.py [hunk 4]
```diff
@@ -84,7 +89,7 @@ def print_key(identity, end=""):
     sign_list = set({})
```
Comment (sourabhtk37) [inline src/gpg/fn/print_key.py]: `set()`

File: src/gpg/key/gen.py [hunk 5]
```diff
@@ -4,23 +4,23 @@
...
 def gen(parameters):
-    params = """<GnupgKeyParms format="internal">
-    {parameters}
-    </GnupgKeyParms>
-    """.format(parameters=parameters)
+    try:
+        params = """<GnupgKeyParms format="internal">
+        {parameters}
+        </GnupgKeyParms>
+        """.format(parameters=parameters)
```
Comment (sourabhtk37) [inline src/gpg/key/gen.py]: Try block should start from line 13 imo i.e Piece of code that are likely to cause exceptions.

File: src/gpg/key/gen.py [hunk 6]
```diff
@@ -4,23 +4,23 @@
 
 
 def gen(parameters):
```
Comment (sourabhtk37) [inline src/gpg/key/gen.py]: A bit more descriptive name helps in readability. `generate_key` or `gen_key`.

**Review by sourabhtk37 — COMMENTED** [review]
Great work @diveshuttam

**Comment by diveshuttam** [comment]
I researched a bit and found about [decorators](https://www.python.org/dev/peps/pep-0318/). I have created handle_exception as a decorator. It wraps both try, except as well as raising/printing error functionality. It shortens code greatly and also looks prettier IMO. @sourabhtk37, @dashohoxha need your suggestions.

**Comment by dashohoxha** [comment]
The definition of the decorator looks like magic to me. The rest looks great.

By the way, you have missed an 'i' in the spelling of 'auxiliary'.

**Comment by diveshuttam** [comment]
>The definition of the decorator looks like magic to me. 

I have added a comment about the decorator in the code.
Someone familiar with python must not have much problem. But still it helps if one understands code.

BTW just for reference:

```python
@decorator
def function():
    something
```

is equivalent to 
```python
def function():
    something

function = decorator(function)
```

I have used two levels of decoration in my case as I am passing exceptions.

> By the way, you have missed an 'i' in the spelling of 'auxiliary'.

Corrected this.

File: src/gpg/key/gen.py (hunk 5 above)
Comment (diveshuttam) [reply src/gpg/key/gen.py]: Yup but it doesn't change the behavior if we have correct lines in `try`. Also am am catching specific exceptions now so it won't mask any unknown exceptions.

**Review by dashohoxha — APPROVED** [review]
It looks OK to me.

**Comment by diveshuttam** [comment]
@sourabhtk37 please go through this once again before merging. Since this introduces a major change.

**Comment by dashohoxha** [comment]
> @sourabhtk37 please go through this once again before merging. Since this introduces a major change.

@diveshuttam it is important that you run the tests and make sure that the changes don't break them.
The things that me and Sourabh can ask for changing are basically small things and not so important. I mean it is not so important whether you write **if** with parentheses or without parentheses, but it is very important if there is a bug.

**Comment by diveshuttam** [comment]
Yes I ran the tests and they are passing. Just asking if the my methodology is right and maybe if @sourabhtk37 has something better in mind.

**Comment by sourabhtk37** [comment]
@diveshuttam The decorators approach is good but putting the entire function code inside a try-except block is not. It might cause unexpected behaviour being python. Just my 2 cents.

**Comment by dashohoxha** [comment]
> @diveshuttam The decorators approach is good but putting the entire function code inside a try-except block is not. It might cause unexpected behaviour being python. Just my 2 cents.

It is not obvious to me what problems it may cause. So, let's keep as it is for the time being, and we can always fix it later in case we notice some problems.

**Comment by diveshuttam** [comment]
>@diveshuttam The decorators approach is good but putting the entire function code inside a try-except block is not. It might cause unexpected behaviour being python. Just my 2 cents.

I have merged this for now. Though the entire function has `try except`, it will only exit as soon as first exception arises, so I think it shouldn't cause much problem. In case I wanted to do something else, and not exit, we can use inner `try except` also as the error binds to the closest `try except` first. I have 
 done similar thing in couple of places.

**Pull request merged**